File: HW_2/problem_1.cpp
#include <iostream>
#include <vector>
using namespace std;
class Stack
{
    /*
    this class is implementation of Stack data strcutre 
    which is LIFO (last in, first out) type data structure
    think of it like you have pile of dishes
    the new dish comes in the top of the pile
    and that new dish will be the first to be washed.
    */
    private:
        // used to store items of the stack
        vector<int> data;
    public:
        Stack();    // default constructor
        Stack(int);
        ~Stack();   // destructor
        void push(int);      // add new element to the stack (ie add new dish)
        int pop();          // move the top element (ie wash the most recent dish)
        int peek();        // return the top element in the stack
        void traverse();  // traverse the stack
        void size();     // returns the number of items in the stack (ie how many dished we have)
};
// end of class definition

Stack::Stack()
{
    this->data.reserve(5);
}

Stack::Stack(int n)
{
    vector<int> data;
    this->data.reserve(n);

}
//==============================
void Stack::push(int item)
{
    // adds the parameter item into top of the stack
    this->data.push_back(item);
}
//==================
int Stack::pop()
{
    // pops and returns the topmost item in the stack
    if (this->data.empty())
    {
        cout <<"ERROR!!! can't remove any element, the stack is empty\n";
        throw -1;
    }
    int temp = this->data.back();
    this->data.pop_back();
    return temp;
}
//======================
int Stack::peek()
{
    // returns topmost item in the stack
    if (this->data.empty())
    {
        cout <<"ERROR!!! can't Peek, the stack is empty\n";
        throw -1;
    }
    return this->data.back();
}
//====================
void Stack::traverse()
{
    if (this->data.empty())
    {
        cout <<"Can't traverse empty stack\n";
        return;
    }
    cout <<"The items in the stack: \n";
    for (auto i = this->data.rbegin(); i != this->data.rend(); ++i)
        cout << *i <<" ";
    cout <<"\n";
}
//=================================
void Stack::size()
{
    cout <<"Total items in the stack: " <<this->data.size() <<endl;
}
//==============================


Stack::~Stack() { }//Destructor
//==================================================================
//end of class implementation

int main()
{
    Stack st;
    st.traverse();
    st.push(3);
    st.push(55);
    st.push(47);
    st.size();
    st.traverse();
    cout <<"Poped: "<<st.pop() <<endl;
    cout << "Peeking: " <<  st.peek() <<endl;
    cout <<"Poped: " << st.pop() <<endl;
    cout << "Peeking: " <<  st.peek() <<endl;

}

File: HW_2/problem_1.py
class Stack:
    """"
    this is a class for Stack data structure.
    Stack is LIFO datastructure (ie Last element in, is the first one removed)
    think of it like you have pile of dishes
    the new dish comes in the top of the pile
    and that new dish will be the first to be washed.
    """
    # initializer
    def __init__(self) -> None:
        self.data = []
    
    # add new item to the stack
    def push(self, item):
        self.data.append(item)
    # remove the recently added item from the stack
    def pop(self):
        if (len(self.data)== 0):
            print("Didn't remove any element, the stack is empty")
            return 
        removed_item = self.data.pop()
        return removed_item
    # returns the most recent element (without removing it)
    def peek(self):
        if (len(self.data) == 0):
            print("Can't peek, the stack is empty")
            return
        return self.data[-1]


#testing the class
if __name__ == "__main__":
    stack = Stack()
    stack.pop()
    stack.peek()
    stack.push(3)
    print("Peeked: ", stack.peek())
    stack.push(6)
    stack.push(9)
    print("Peeked: ", stack.peek())
    print("data removed: ", stack.pop())
    print("peeked: ", stack.peek())
